pkg/collector/containerlogs_collector_containerd.go

```diff
@@ -0,0 +1,167 @@
+		}
+	}
+	return selectedContainerLogs
+}
+
+//DoesSelectorSelectContainerLog contains the logic for determining if a selector selects a containerLog for collecting
+func (collector *ContainerLogsCollectorContainerD) DoesSelectorSelectContainerLog(containerLog ContainerLog, selector ContainerLogSelector) bool {
+	return containerLog.namespace == selector.namespace && strings.HasPrefix(containerLog.containerName, selector.containerNamePrefix)
+}
+
+//ParseContainerLogSelectors parses selectorStrings into component struct
+//TODO allow the raw struct objects to be defined directly in the deployment yaml and add additional required logic to DoesSelectorSelectContainerLog
+func (collector *ContainerLogsCollectorContainerD) ParseContainerLogSelectors(selectorStrings []string) []ContainerLogSelector {
+	var containerLogSelectors []ContainerLogSelector
+
+	for _, selectorString := range selectorStrings {
+		selectorStringParts := strings.Split(selectorString, "/")
+
+		if len(selectorStringParts) == 1 {
+			containerLogSelectors = append(containerLogSelectors, ContainerLogSelector{
+				namespace: selectorStringParts[0],
+			})
+		}
+		if len(selectorStringParts) == 2 {
+			containerLogSelectors = append(containerLogSelectors, ContainerLogSelector{
+				namespace:           selectorStringParts[0],
+				containerNamePrefix: selectorStringParts[1],
+			})
+		}
+	}
+
+	return containerLogSelectors
+}
+
+//ParseContainerLogFilenames parses container log filenames into component struct
+func (collector *ContainerLogsCollectorContainerD) ParseContainerLogFilenames(directoryPath string, containerLogFilenames []string) []ContainerLog {
+
+	var containerLogs []ContainerLog
+
+	//container log files are in format PODNAME_NAMESPACE_CONTAINERNAME-CONTAINERUID.log
+	//TODO check that CONTAINERNAME and CONTAINERUID are correct terminology
+	for _, logFile := range containerLogFilenames {
+		logfileSplitOnDot := strings.Split(logFile, ".")
+		logFileSplitOnUnderscore := strings.Split(logfileSplitOnDot[0], "_")
+		containerNameWithIDSplitOnDash := strings.Split(logFileSplitOnUnderscore[2], "-")
+
+		//uid is the last value
+		indexOfUid := len(containerNameWithIDSplitOnDash) - 1
+
+		//containerName is everything except the last value, joined
+		containerName := strings.Join(containerNameWithIDSplitOnDash[0:indexOfUid], "")
+
+		containerLogs = append(containerLogs, ContainerLog{
+			podname:       logFileSplitOnUnderscore[0],
+			namespace:     logFileSplitOnUnderscore[1],
+			containerName: containerName,
+			containeruid:  containerNameWithIDSplitOnDash[indexOfUid],
+			filepath:      path.Join(directoryPath, logFile),
+		})
+	}
+
+	return containerLogs
+}
+
+//GetAllContainerLogFilesThatHaveEverRunOnHost gets the list of log files for all containers that have ever run on the host
+func (collector *ContainerLogsCollectorContainerD) GetAllContainerLogFilesThatHaveEverRunOnHost() ([]string, error) {
+	output, err := utils.RunCommandOnHost("ls", containerLogDirectory)
```

Comment:
We should address removal of RunCommandOnHost all at once and as separate PR. Similar with windows which isn't currently supported anyway

Reply:
The only reason it's not supported is this runcmd so let's at least limit the damage: we could leave legacy code as is but any new code shall not rely on these methods

Reply:
Do you have ideas or suggestions for how to replace RunCommandOnHost? We can't call code that uses it legacy without a concrete plan for how to replace it, and I haven't myself come across anything suitable, or even confirmed that it an alternative is feasible.

Being able to run commands in the namespace of the host is one of the core functions of Periscope, and so "just this one function" is a big blocker - until this is clarified it seems unreasonable to halt any PRs on the basis that they leverage this method.

Reply:
Yeah the idea is to use adequate libraries : for kubectl use https://github.com/kubernetes/client-go, for helm use https://github.com/helm/helm/blob/main/pkg/kube/client.go, to access files we can mount volumes and use io / ioutils libraries, etc ...
I'm not saying to remove RunCmdOnHost now but aks-periscope is not aimed to be a wraper to call cli tools. We have libraries to do it properly so let's do it :)
Leave old code as is but for new code, let's use the libraries, especially regarding files

Reply:
Ahhh right ok will try with volume mount, thanks

Reply:
FYI the change to use a volume mount has forced a refactor to target the logs in /var/log/pods/ not /var/log/containers/, as /containers/ just contains symlinks to the actual log files in /pods/ :doh:

Reply:
This is the reason why we mount /var/log:
```yaml
 volumes:
      - name: varlog
        hostPath:
          path: /var/log
```

Reply:
yes, however we need to change the code significantly to read from /var/log/pods rather than /var/log/containers as the directory structure is different. With /var/log mounted it still wouldn't follow the symlink from /containers to /pods